### PR TITLE
Memprof support for unmarshalled data

### PR DESCRIPTION
**Pull request by jhjourdan**
This PR implements memprof tracking for unmarshaled data.

This is performed by traversing the unmarshaled section after unmarshalling is complete. The reasons for not calling the memprof tracking function during unmarshaling are:
- The current code is less invasive in the intern.c code.
- In the current code, there is no performance penalty when memprof is disabled (while some code would have to be executed at every unmarshalled block otherwise)
- Benchmarking suggests that even when memprof is enabled, re-traversing the unmarshaled section is faster. Perhaps the reason in that the hot part of the re-traversing algorithm is a very tight loop.

In addition to this, this PR implements memprof's own root tracking mechanism for the postponed queue. Calling the `caml_xxx_global_root` functions turned out to be rather slow.

TODO:
- [x] Write tests

### CODE REVIEW

**Comment by gasche** [comment]
This stayed on my "pile of things to look at urgently" for about a week now and I haven't had the time to look, nor will I have in the closure future. Sorry for the non-progress report. Writing tests definitely sounds like a good idea to me, though.

**Comment by stedolan** [comment]
It's also been on my pile, along with responding to your comments on #8731! I'll have time to do both on Monday, but not before.

**Comment by jhjourdan** [comment]
No worries, I am also rather busy these days.

File: runtime/memprof.c
Comment (stedolan) [inline runtime/memprof.c]: Introducing `szof` for `sizeof (struct postponed_block)` makes this harder to read.

File: runtime/memprof.c
Comment (stedolan) [inline runtime/memprof.c]: Is this right? It seems to include the uninitialised empty slot at `postponed_hd` in `sz1`, which gets copied below.

File: runtime/memprof.c
Comment (stedolan) [inline runtime/memprof.c:299]: What's the motivation for reimplementing this resize code? The new version's harder to read: the old one with the while loop was obviously correct, and I'm not sure whether the new version is. Is it for performance? How often do you expect this to run?

File: runtime/memprof.c
Comment (stedolan) [inline runtime/memprof.c]: This is a bit hard to read.

File: runtime/intern.c
Comment (stedolan) [inline runtime/intern.c:680]: This OOM check is new. Was this just wrong before?

**Review by stedolan — CHANGES_REQUESTED** [review]
Looks good! (This is a much smaller patch than I'd thought: I didn't realise that the regenerated `.depend` made up most of the changed lines)

I'm not completely sold on the mechanism for sharing stacks between samples within a single unmarshalling block. Making statmemprof go faster on unmarshalling allocations than on other allocations doesn't seem very important to me, because unmarshalling allocations are an uncommon case. (I'll do some benchmarking of this when I'm fixing up #8731).

File: runtime/intern.c
Comment (jhjourdan) [reply runtime/intern.c:680]: Indeed.

Comment (jhjourdan) [reply runtime/intern.c:680]: Do you think this deserves its own PR? This bug is quite difficult to trigger...

File: runtime/memprof.c
Comment (jhjourdan) [reply runtime/memprof.c:299]: I had the feeling that the `memcpy` code was more natural (eh, we are just moving some data, after all!), and yes, performance is better.

But I agree the end result is not so good. I reverted the initial code.

**Comment by jhjourdan** [comment]
The flambda Travis is doing a timeout. Looking at the log, it does not seem to be an infinite loop. Also, this timeout seems reproducible, since it has happened for each of the pushes of this PR.

Usually the Travis CI for flambda lasts ~30min, and the timeout is set at 50min. So, it seems that this PR created a massive slowdown. I'm investigating.

**Comment by jhjourdan** [comment]
The reason of the slowdown seems to be what I described in #8893. This is fixed.

I wrote some tests, rebased, squashed small commits and included @stedolan comments. As far as I am concerned, this PR is ready to merge.

File: runtime/intern.c
Comment (stedolan) [reply runtime/intern.c:680]: No, I think it's fine as part of this one.

**Review by stedolan — APPROVED** [review]

**Comment by stedolan** [comment]
This looks good to me! Good to merge once CI passes (which will need some `make depend` & merge conflict fixing)

**Comment by jhjourdan** [comment]
The conflicts are now solved, so we are just waiting for CI to succeed before merging.

**Comment by jhjourdan** [comment]
@gasche : could you merge this? @stedolan approved it, and I'm afraid of conflicts if we wait he comes back from vacations.

File: runtime/intern.c
Comment (gasche) [inline runtime/intern.c:810]: Note: I think it would be nicer to have the same level of abstraction in the success and failure cases, with two `intern_end_{success,failure}` functions .

**Pull request merged**
